### PR TITLE
fix: add guard against NodeStatus

### PR DESCRIPTION
**Pull request by isubasinghe**
Fixes #TODO

Please do not open a pull request until you have checked ALL of these:

* [ ] Create the PR as draft .
* [ ] Run `make pre-commit -B` to fix codegen and lint problems.
* [ ] Sign-off your commits (otherwise the DCO check will fail).
* [ ] Use [a conventional commit message](https://www.conventionalcommits.org/en/v1.0.0/) (otherwise the commit message check will fail).
* [ ] "Fixes #" is in both the PR title (for release notes) and this description (to automatically link and close the issue).
* [ ] Add unit or e2e tests. Say how you tested your changes. If you changed the UI, attach screenshots.
* [ ] Github checks are green.
* [ ] Once required tests have passed, mark your PR "Ready for review".

If changes were requested, and you've made them, dismiss the review to get it reviewed again.

### CODE REVIEW

**Comment by isubasinghe** [comment]
# POST MORTEM 

Nodes are defined as a map[string]NodeStatus. This combined with zero values returns resulted in the bug our customer encountered of container sets causing random failure.  This confirms my (later stated) hunch, it's actually not the implementation of container sets that is causing issues but interactions of various components that weren't anticipated. 

Unfortunately this brings to the point I had made to @JPZ13 earlier, making changes to Argo Workflows does requires a near 1 to 1 mapping of the Workflows codebase into memory, at the size of the codebase right now, that is an impossible task for most (but almost certainly everyone involved in the project).  

To add to this as an example, I encountered a case where a bug prevented a bug from going through in the first place, that is they cancelled each other out, if you look at [here](https://github.com/pipekit/argo-workflows/pull/22/files#diff-1b23041c82907739e08d36ec5cc401a4f54e88dd12f9d37d29880da7ce6672acR839-R840), there was no mention of what to do if node was not present in the hashmap, but an invalid access resulted in an empty struct, this meant the boolean variables here get changed correctly.  

This is to me fairly terrifying because it is impossible to distinguish between bug and intentional behaviour without knowing what the original author intended. I think the code should be as obvious as possible, when an explicit check is made against a map, the author also asserts the intended behaviour explicitly. 

To generalise on the above comments, the core problem is that the codebase isn't as modular as it really should be, different sections of the codebase assume behaviour from other sections of the codebase. This is what makes the codebase hard to debug, hard to add features to and hard to change, ultimately I believe the container set issues were only just a symptom of the state of the codebase, the core issues lie within Argo Workflows itself. Putting time and effort into simplifying Workflows as much as possible will prevent future "whack-a-mole" type situations.   


# What slowed me down
I ended up chasing a wild goose of sorts, I assumed the error originated from boundaryID being "". This is actually an intended value for some nodes, but given zero values are a thing in Go, it is impossible to distinguish between:
-  "is this accidental because the Nodes map returned a zero value struct?"
- "is this accidental because somewhere boundaryID was set to an empty string?", 
- "is this intentional because for some nodes, the boundaryID can be an empty string?". 

In reality the actual strongly typed type of boundaryID is really `Option[string]` not `string`.  This forms the basis of my argument against using zero values, it is impossible to distinguish what they really mean. 

# Prevention of this particular type of issue
1. Single Get/Set functions so that tracing issues relating to them can be done via debugging one function. 
2. Avoid relying on zero values completely. This is somewhat anti-go but frankly I think this is a flaw in the language design. Relying on zero values doesn't scale well (with respect  to features/code/(number of engineers) not performance). 
3. Prefer panicking over zero values, a panic is an easy bug to fix, chasing down code that other people wrote that may or may not be the reason your workflow fails is really not. 
4. Write a simple static analysis tool that detects accessing of map that have a non pointer return type. Essentially ensure that all maps are of the form map[A]*B not map[A]B. An access on a nil pointer is a panic, this should be the preferred option.

File: server/artifacts/artifact_server.go [hunk 1]
```diff
@@ -10,6 +10,7 @@ import (
 	"path"
 	"strings"
 
+	"github.com/sirupsen/logrus"
```
Comment (Joibel) [inline server/artifacts/artifact_server.go]: logrus is already imported as log on the next line

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 2]
```diff
@@ -22,6 +22,8 @@ import (
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/apimachinery/pkg/util/wait"
 
+	"github.com/sirupsen/logrus"
```
Comment (Joibel) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: logrus is always imported as log elsewhere in the codebase
`log "github.com/sirupsen/logrus"`

File: util/resource/updater.go [hunk 3]
```diff
@@ -1,6 +1,8 @@
 package resource
 
 import (
+	"github.com/sirupsen/logrus"
```
Comment (Joibel) [inline util/resource/updater.go]: import as log

File: workflow/controller/pod_cleanup.go [hunk 4]
```diff
@@ -23,7 +25,12 @@ func (woc *wfOperationCtx) queuePodsForCleanup() {
 			continue
 		}
 		nodeID := woc.nodeID(pod)
-		if !woc.wf.Status.Nodes[nodeID].Phase.Fulfilled() {
+		nodePhase, err := woc.wf.Status.Nodes.GetPhase(nodeID)
+		if err != nil {
+			woc.log.Fatalf("was unable to obtain node for %s", nodeID)
+			panic(fmt.Sprintf("node was not present with id %s", nodeID))
```
Comment (Joibel) [inline workflow/controller/pod_cleanup.go]: I'd suggest we shouldn't abandon looking through all the other pods here, and should complain loudly but `continue`

File: workflow/controller/steps.go [hunk 5]
```diff
@@ -263,50 +298,61 @@ func (woc *wfOperationCtx) executeStepGroup(ctx context.Context, stepGroup []wfv
...
+	if err != nil {
+		return nil, err
+	}
 	// Return if not all children completed
 	completed := true
 	for _, childNodeID := range node.Children {
-		childNode := woc.wf.Status.Nodes[childNodeID]
+		childNode, err := woc.wf.Status.Nodes.Get(childNodeID)
+		if err != nil {
+			woc.log.Fatalf("was unable to obtain node for %s", childNodeID)
```
Comment (Joibel) [inline workflow/controller/steps.go]: nit: these two messages should be identical?

File: workflow/controller/taskresult.go [hunk 6]
```diff
@@ -55,25 +56,29 @@ func (woc *wfOperationCtx) taskResultReconciliation() {
 	for _, obj := range objs {
 		result := obj.(*wfv1.WorkflowTaskResult)
 		nodeID := result.Name
-		old := woc.wf.Status.Nodes[nodeID]
-		new := old.DeepCopy()
+		old, err := woc.wf.Status.Nodes.Get(nodeID)
+		if err != nil {
+			woc.log.Fatalf("was unable to obtain node for %s", nodeID)
+			panic(fmt.Sprintf("unable to obtain node for %s", nodeID))
```
Comment (Joibel) [inline workflow/controller/taskresult.go]: panic seems extreme here, continue instead?

**Review by Joibel — CHANGES_REQUESTED** [review]
I haven't commented on every use of logrus, but they should all change.

File: workflow/controller/taskresult.go (hunk 6 above)
Comment (isubasinghe) [reply workflow/controller/taskresult.go]: I bubbled up an error instead, I think this is safer than continuing.

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 7]
```diff
@@ -1704,6 +1706,57 @@ func (n Nodes) Find(f func(NodeStatus) bool) *NodeStatus {
 	return nil
 }
 
+func (n Nodes) Get(key string) (*NodeStatus, error) {
```
Comment (JPZ13) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1711]: nit: Add comments for exported functions

File: pkg/apis/workflow/v1alpha1/workflow_types.go [hunk 8]
```diff
@@ -1704,6 +1706,57 @@ func (n Nodes) Find(f func(NodeStatus) bool) *NodeStatus {
...
+	return &val, nil
+}
+
+func (n Nodes) Has(key string) bool {
+	_, err := n.Get(key)
+	return err == nil
+}
+
+func (n Nodes) GetPhase(key string) (*NodePhase, error) {
+	val, ok := n[key]
```
Comment (JPZ13) [inline pkg/apis/workflow/v1alpha1/workflow_types.go]: nit: You can use the `Get` function you implemented here for this and the other times you do `n[key]`

**Review by Joibel — APPROVED** [review]

**Pull request merged**